### PR TITLE
feat: Implement Tracking in .NET #309

### LINKED ISSUE #309: [FEATURE] Implement Tracking
### Requirements

_No response_

### PR DESCRIPTION
**Pull request by chrfwow**
## This PR
Adds support for tracking 

### Related Issues

Closes #309

### CODE REVIEW

File: src/OpenFeature/Model/TrackingEventDetails.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+/// <summary>
+/// The `tracking event details` structure defines optional data pertinent to a particular `tracking event`.
+/// </summary>
+/// <seealso href="https://github.com/open-feature/spec/blob/main/specification/sections/06-tracking.md#62-tracking-event-details"/>
+public sealed class TrackingEventDetails
+{
+    /// <summary>
+    /// The index for the "targeting key" property when the EvaluationContext is serialized or expressed as a dictionary.
+    /// </summary>
+    internal const string TargetingKeyIndex = "targetingKey";
```
Comment (chrfwow) [inline src/OpenFeature/Model/TrackingEventDetails.cs:16]: Do we need the `TargetingKeyIndex` here?

Comment (toddbaert) [reply src/OpenFeature/Model/TrackingEventDetails.cs:16]: No, we would need instead a `value` index if anything, that would serve the same purpose when the TrackingEventDetails are serialized.

File: src/OpenFeature/FeatureProvider.cs [hunk 2]
```diff
@@ -140,5 +141,16 @@ public virtual Task ShutdownAsync(CancellationToken cancellationToken = default)
         /// </summary>
         /// <returns>The event channel of the provider</returns>
         public virtual Channel<object> GetEventChannel() => this.EventChannel;
+
+        /// <summary>
+        /// Use this method to track user interactions and the application state. The implementation of this method is optional.
```
Comment (toddbaert) [inline src/OpenFeature/FeatureProvider.cs:146]:
```suggestion
        /// Track a user action or application state, usually representing a business objective or outcome. The implementation of this method is optional.

```

File: test/OpenFeature.Tests/OpenFeatureClientTests.cs [hunk 3]
```diff
@@ -540,5 +540,101 @@ public void ToFlagEvaluationDetails_Should_Convert_All_Properties()
...
+        {
+            const string key = "key";
+            const string global = "global";
+            const string client = "client";
+            const string invocation = "invocation";
+            var globalEvaluationContext = new[] { new EvaluationContextBuilder().Set(key, "global").Build(), null };
+            var clientEvaluationContext = new[] { new EvaluationContextBuilder().Set(key, "client").Build(), null };
+            var invocationEvaluationContext = new[] { new EvaluationContextBuilder().Set(key, "invocation").Build(), null };
+
+            var data = new TheoryData<string, EvaluationContext?, EvaluationContext?, EvaluationContext?, string>();
```
Comment (toddbaert) [inline test/OpenFeature.Tests/OpenFeatureClientTests.cs:598]: Took me a minute to wrap my head around this but it makes sense! :sweat_smile:

File: src/OpenFeature/Model/TrackingEventDetails.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    /// <summary>
+    ///A predefined value field for the tracking details.
+    /// </summary>
+    public readonly double? Value;
```
Comment (toddbaert) [inline src/OpenFeature/Model/TrackingEventDetails.cs:21]: Optional is good.

File: README.md [hunk 5]
```diff
@@ -74,6 +74,7 @@ public async Task Example()
 | ✅      | [Providers](#providers) | Integrate with a commercial, open source, or in-house feature management tool.                                                     |
 | ✅      | [Targeting](#targeting) | Contextually-aware flag evaluation using [evaluation context](https://openfeature.dev/docs/reference/concepts/evaluation-context). |
 | ✅      | [Hooks](#hooks)         | Add functionality to various stages of the flag evaluation life-cycle.                                                             |
+| ✅      | [Tracking](#tracking)   | Associate user actions with feature flag evaluations.                                                                              |
```
Comment (toddbaert) [inline README.md:77]: I couldn't suggest this so I just added it - hope you don't mind. This is useful here but it's also used to build [this](https://openfeature.dev/docs/reference/technologies/sdk-compatibility).

**Review by toddbaert — CHANGES_REQUESTED** [review]
I think [this](https://github.com/open-feature/dotnet-sdk/pull/327#discussion_r1873854632) is the only blocker.

File: test/OpenFeature.Tests/OpenFeatureClientTests.cs (hunk 3 above)
Comment (chrfwow) [reply test/OpenFeature.Tests/OpenFeatureClientTests.cs:598]: Should I add some documentation to it?

**Review by askpt — COMMENTED** [review]
I added a couple of concerns I found during the review. Otherwise, it seems good!

File: src/OpenFeature/OpenFeatureClient.cs [hunk 6]
```diff
@@ -367,6 +367,31 @@ private async Task TriggerFinallyHooksAsync<T>(IReadOnlyList<Hook> hooks, HookCo
...
+        /// <summary>
+        /// Use this method to track user interactions and the application state.
+        /// </summary>
+        /// <param name="trackingEventName">The name associated with this tracking event</param>
+        /// <param name="evaluationContext">The evaluation context used in the evaluation of the flag (optional)</param>
+        /// <param name="trackingEventDetails">Data pertinent to the tracking event (Optional)</param>
+        /// <exception cref="ArgumentException">When trackingEventName is null or empty</exception>
+        public void Track(string trackingEventName, EvaluationContext? evaluationContext = default, TrackingEventDetails? trackingEventDetails = default)
+        {
+            if (string.IsNullOrEmpty(trackingEventName))
```
[the remark this answers is not in the document]
Comment (chrfwow) [reply src/OpenFeature/OpenFeatureClient.cs:379]: At least in the [Java SDK](https://github.com/open-feature/java-sdk/pull/1228/files#diff-e0fc1e23041e08fa8dd72864acce5b949f84987f405accd4935f233de281ef14R232), we do allow empy strings

[the remark this answers is not in the document]
Comment (toddbaert) [reply src/OpenFeature/OpenFeatureClient.cs:379]: I think we should not allow the white space.

File: README.md [hunk 7]
```diff
@@ -212,6 +213,19 @@ await Api.Instance.SetProviderAsync(myClient.GetMetadata().Name, provider);
 myClient.AddHandler(ProviderEventTypes.ProviderReady, callback);
 ```
 
+### Tracking
+
+The tracking API allows you to use OpenFeature abstractions and objects to associate user actions with feature flag evaluations.
```
Comment (lukas-reining) [inline README.md:218]:
```suggestion
The [tracking API](https://openfeature.dev/specification/sections/tracking) allows you to use OpenFeature abstractions and objects to associate user actions with feature flag evaluations.
```

Adding a link would be helpful.

**Review by toddbaert — APPROVED** [review]
Approved, but consider @askpt 's points. I think it's fine to throw for whitespace event keys as well.

**Review by lukas-reining — APPROVED** [review]

File: src/OpenFeature/OpenFeatureClient.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrfwow) [reply src/OpenFeature/OpenFeatureClient.cs]: I changed it now to throw on whitespace strings, but keep in mind that we are inconsistent with the Java SDK.

**Review by askpt — APPROVED** [review]

**Review by toddbaert — APPROVED** [review]

**Pull request merged**